src/lib/Inference.hs

```diff
@@ -125,10 +125,105 @@ inferTopUDecl decl@(ULet _ (UPatAnn p ann) rhs) result = do
     applyDefaults
     return val
   return $ UDeclResultWorkRemaining block $ Abs decl result
-inferTopUDecl (UEffectDecl _ _ _) _ = throw NotImplementedErr "inferTopUDecl::UEffectDecl"
-inferTopUDecl (UHandlerDecl _ _ _ _ _ _) _ = throw NotImplementedErr "inferTopUDecl::UHandlerDecl"
+inferTopUDecl (UEffectDecl opTys effName opNames) result = do
+  opTys' <- forM opTys $ \opTy -> liftInfererM $ solveLocal $ inferEffOpType opTy
+  let effSourceName = uBinderSourceName effName
+  let effOpSourceNames = nestToList uBinderSourceName opNames
+  let effDef = EffectDef effSourceName (zip effOpSourceNames opTys')
+  effName' <- emitEffectDef effDef
+  opNames' <-
+    forM (enumerate effOpSourceNames) \(i, opName) -> do
+      emitEffectOpDef effName' i opName
+  let subst = effName @> effName' <.> opNames @@> opNames'
+  UDeclResultDone <$> applySubst subst result
+-- TODO(alex): handle retEff
+inferTopUDecl (UHandlerDecl effName bodyTyArg tyArgs _retEff retTy opDefs handlerName) result =
+  do let (InternalName _ effName') = effName
+     let handlerSourceName = uBinderSourceName handlerName
+     Abs Empty (Abs bodyTyArg' (Abs tyArgs' (retTy' `PairE` retOp' `PairE` ListE ops'))) <-
+      liftInfererM $ solveLocal $ buildDeclsInf do
+        checkHandlerBodyTyArg bodyTyArg \r -> do
+          checkHandlerArgs tyArgs do
+            effName'' <- sinkM effName'
+            retTy' <- checkUType retTy
+            (retOp, ops') <- checkHandlerBody effName'' (sink r) retTy' opDefs
+            return (retTy' `PairE` retOp `PairE` ListE ops')
+     -- TODO(alex): replace Pure with actual effects
+     let handler = HandlerDef effName' bodyTyArg' tyArgs' Pure retTy' ops' retOp'
```

Comment:
What does this `Pure` mean? Is this even sound to merge without this fix?

Reply:
This is unsound for handlers whose operations themselves call effect functions... we're just trying to push through to make `Exn` work and are ignoring this case (for now).